Make `attest-enroll` more configurable, pluggable; make it escrow and make it use AEAD

nicowilliams opened the pull request:
Tested manually.

This PR adds:

 - re-enroll and enroll-more options
 - pluggable/configurable generators of contents for an enrollment -- this will be used to provision an enrolled host with: a signature key and certificate, a host keytab, and maybe other things
 - authenticated symmetric encryption of secrets
 - escrow to zero, one, or more bare RSA public keys
 - escrow to zero, one, or more TPMs (with optional policy)
 - TK or EK method of secrets transport to enrolled host, with optional, configurable policy
 - configuration file (though everything has a command-line option as well)
 - index by hostname

TBD:

 - add commit hook so we can do `git add`/`git commit`/`git push`, or whatever, at the end of enrollment
 - add a test (which should start multiple swTPMs, some for escrow, some for enrollees, and which should test secret recovery by enrollee and by escrowers, policies, etc.)

Current usage message:

```
Usage: attest-enroll [OPTIONS] HOSTNAME < ekpub
       attest-enroll [OPTIONS] -I EKPUB HOSTNAME

  Options:

    -h          This message
    -v          Verbose
    -x          Trace

    -a          Add to existing enrollment of this EKpub
    -r          Replace the current enrollment of this EKpub
    -I EKPUB    EKpub file (default: read from stdin)

  Configuration options:

    -C CONF     Config file (default: /etc/safeboot-enroll.conf)
    -E DIR      Directory containing PEM files containing escrow keys.
                These should be named "tpm-<name>.pub" and contain an
                escrow TPM's EKpub, or "<name>.pem" and contain an RSA
                key for PKCS encryption.
    -G GENPROG  A program to generate a secret and/or metadata (details below).
                May be given multiple times.  Hostname (public) and filesystem
                keys (secret) are always generated first.
    -O DIR      Output directory (default: /safeboot/sbin/build/attest)
    -P POLICY   A policyDigest or a script that outputs one.  The TPM to which
                the enrolled EKpub belongs _will_ enforce the policy when
                decrypting secrets for it.
                (default: )
    -T METHOD   Secrets transport method.  MEDTHOD is one of:
                 - TK (uses a long-lived RSA transport key whose primary key is
                       TPM2_Duplicate()ed to the enrolled EKpub)
                 - EK (uses TPM2_MakeCredential() to encrypt to the enrolled
                       EKpub)
                (default: EK)

  CONF is a bash script that should set/clear one or more of:

    DBDIR ESCROW_PUBS EKPUB POLICY ESCROW_POLICY TRANSPORT_METHOD
    GENPROGS (bash array)

  GENPROGS will be called in the order they are given, with the following
  arguments:

    TMP-OUTPUT-DIR ENROLL-DIR HOSTNAME

  and are expected to a) create a file in OUTPUT-DIR, b) output:

    sensitive NAME1 [NAME2 ..].

  or

    public NAME1 [NAME2 ..].

  or

    skip [REASON]

  E.g.,

    sensitive SOMEPRIVKEY

    sensitive SOMEPRIVKEY.PEM SOMEPUBKEY.PEM SOMECERT.PEM

    public SOME-META-DATA-HERE

    skip NOT NEEDED

  The NAMEs are of files in the TMP-OUTPUT-DIR.  Any sensitive NAME1 file will
  be encrypted and escrowed; all other files will be copied to the enrollment
  area for the given EKpub.
```

nicowilliams commented:
Need to add tests.

nicowilliams commented:
Here's a sample config:

```
: ; cat /etc/safeboot-enroll.conf
DBDIR=/tmp/o
ESCROW_PUBS=/tmp/o/e
TRANSPORT_METHOD=EK
GENPROGS=(genhostname genfskey gentest0 gentest1 gentest2 gentest3 gentest4)
POLICY=
ESCROW_POLICY=
: ;
```

The `gentestN` geprogs are shell functions in `attest-enroll` to test `GENPROGS`/`-G` functionality.

Each `genprog` can:

 - do nothing (output `skip`)
 - output zero or one secret file (to be encrypted to the enrolled device and to escrow keys)
 - output zero, one, or more public metadata files (not encrypted nor escrowed)
 - specify whether the outputs are to be unique or qualified and deduplicated

Here are the test `genprog`s:

```bash
gentest0() {
    dd if=/dev/urandom of=${1}/test0 bs=16 count=1 2>/dev/null
    echo test0pub > "${1}/test0pub"
    echo "sensitive unique test0 test0pub"
}

gentest1() {
    echo "${USER:-test1}" > "${1}/test1"
    echo test1pub > "${1}/test1pub"
    echo "sensitive idempotent test1 test1pub"
}

gentest2() {
    echo test2 > "${1}/test2"
    echo test2pub > "${1}/test2pub"
    echo "public unique test2 test2pub"
}

gentest3() {
    echo test3 > "${1}/test3"
    echo test3pub > "${1}/test3pub"
    echo "public idempotent test3 test3pub"
}

gentest4() {
    echo test4$((RANDOM)) > "${1}/test4"
    echo test4pub > "${1}/test4pub"
    echo "public idempotent test4 test4pub"
}
```

`unique` means that there should be just one set of the named files in the DB directory for the given `EKpub`.

`idempotent` means that there may be any number of files for this `genprog` in the DB directory for the given `EKpub`, but to dedup based on the content of the first listed file.

Note that multiple hostnames are allowed, if you keep enrolling the same `EKpub` with different hostnames.  If using a TK, the same TK will be used in all cases.

Next up I'll add `genprog`s to:

 - generate a private signing key and a certificate for its public key, with the hostname as a `dNSName` SAN, say
 - generate a certificate for the `EKpub`
 - extract host keytabs for Kerberos for the hostname

nicowilliams commented:
Maybe this script could also `git add`, `git commit`, and `git push`?

nicowilliams commented:
TBD: Use authenticated symmetric encryption.

nicowilliams commented:
I think I'll get rid of the unique/dedup stuff.  The `genprog`s can do that themselves.

nicowilliams commented:
Now with a test (which I'm running manually).  I need to integrate it into the Mariner and non-Mariner workflows.

The test starts an swTPM for each of three clients and an swTPM for an escrow agent, checks that all the clients can enroll, and then checks that `tpm-receive.sh` can be used to decrypt each client's secrets with the corresponding client's swTPM, and that the escrow agent swTPM can decrypt the escrowed secrets.

TBD:

 - run `tpm2-attest` and friends to check that the simulated clients can attest
 - check that each client can only decrypt its own secrets
 - test that policies work

nicowilliams commented:
Not ready.  I've got to finish changing various bits of various scripts.

nicowilliams commented:
I may need to make each `genprog` do its own escrow and encryption to TPM.  Why?  Well:

 - first, it may not be possible to pass the other secrets to the post mount of / boot sequence (??)
 - second, so that we can have different policies for the `rootfs.key` and other things -- that way the policy for the `rootfs.key` can be that PCR 11 is unextended, and the policy for post early boot things can be that PCR 11 has been extended with a particular value.

Thoughts?

nicowilliams commented:
I now have a complete test -minus HTTP- that works.  It tests `attest-enroll`, `tpm2-attest quote`, `tpm2-attest verify`, `attest-verify`, `tpm2-attest seal`, `tpm2-attest unseal`, and `initramfs/bootscript` (the last one with an env var that makes it exit after recovering the `rootfs.key` file).

I've also added TOFU golden PCR learning, which I needed for now, especially for testing.

I've removed uses of `tar`'s `-z`, but I've not done `s/\<tgz\>/tar/g` yet.

TBD:

 - integrate with @geoffthorpe's flask and mariner build and test framework
    - there's a bit of work to do around enrollment
 - add support for different policies for earliest boot and later early boot (which we need)
   (the idea being that we could have the unextended-PCR#11 policy for `rootfs.key` and extended-once-PCR#11 policy for other secrets, that way we can attest twice, once during initramfs, and once after, and not have to worry about how to get other secrets from initramfs to post pivot)
 - add scripts for keytabs and such

nicowilliams commented:
Regarding privsep... the code that does the secrets generation, and the git add/commit/push, all has to be privsep'ed from the flask code (HTTP server) if we're going to do privsep at all, because otherwise a compromise of the HTTP server would allow the attacker to push known-to-them `rootfs.key`s for not-yet-installed systems, and then the attacker would know those systems' root keys.

osresearch commented:
"77 files changed".  This is going to take a little while to review...

For the most part the fixes for more modern `bash` look good, especially the ones around `local` variable scoping so that we don't have accidental scope creep.

Is the `tar -zcf` to `tar -cf` to remove a dependency on `gzip` or is there another reason that I'm missing?

The different policies for different keys makes lots of sense.  That's something that we should figure out if there is a clean way to specify the alternate policy.

More later.

osresearch commented:
Regarding the changes to `die()`, I agree that having a function inside a library call `exit` is not the best approach and can make error handling harder.  Although having `die()` *not* exit means that the name is no longer as appropriate as the [Perl-inspired `die` function](https://perldoc.perl.org/functions/die).   My suggestion is that we split this into five bash functions:

* `die`: print a message to stderr and exit with an error code. This is only to be used in extreme emergencies.
* `error`: print a message to stderr and return an error code. This is used by most functions when they encounter an error, replacing most of the current uses of `die` and often with the pattern: `foo || error "error!" || return 1`
* `warn`: print a message to stderr when something isn't quite right, but we're going to keep on trying
* `info`: print a message to stderr to update the user on progress (or invoke gnu texinfo, but nobody does that)
* `debug`: print a message to stderr if a `$DEBUG` environment variable is set

osresearch commented:
The house style is one hard-tab for each indentation and for multi-line shell commands to put the arguments one tab stop deeper, regardless of the length of the shell command so that changing commands doesn't require re-indenting the additional lines, with one argument and parameter per line, and with the line continuation characters one space separated from the end.

For tpm2 subcommands, I've been meaning to make sure that we're using long arguments everywhere to try to document their meaning as well. Not all the code has been updated to use this, I'm trying to do it consistently in the new code.

So instead of this:

```
	tpm2 loadexternal -C n -Gecc -r "${TMPDIR}/wkpriv.pem"	\
			  -L "${TMPDIR}/activate-policy.dat"	\
			  -a 'sign|adminwithpolicy'		\
			  -c "${TMPDIR}/wk.ctx"
```

The style guide would call for:
```
	tpm2 loadexternal \
		--hierarchy 'null' \
		--key-algorithm 'ecc' \
		--private "${TMPDIR}/wkpriv.pem" \
		--policy "${TMPDIR}/activate-policy.dat" \
		--attributes 'sign|adminwithpolicy' \
		--key-context "${TMPDIR}/wk.ctx"
```

osresearch commented:
For `sbin/attest-enroll`, shellcheck has some concerns:
```
In sbin/attest-enroll line 34:
. "$BASEDIR/../functions.sh"
  ^------------------------^ SC1090: Can't follow non-constant source. Use a directive to specify location.


In sbin/attest-enroll line 142:
	exit ${1:-1}
             ^-----^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
	exit "${1:-1}"


In sbin/attest-enroll line 153:
G) GENPROGS+=($OPTARG);;
              ^-----^ SC2206: Quote to prevent word splitting/globbing, or split robustly with mapfile or read -a.


In sbin/attest-enroll line 172:
[[ -n ${CONF} && -f ${CONF} ]] && . "${CONF}"
                                    ^-------^ SC1090: Can't follow non-constant source. Use a directive to specify location.


In sbin/attest-enroll line 206:
    ((VERBOSE)) && echo info: "$@" 1>&2 || true
                ^-- SC2015: Note that A && B || C is not if-then-else. C may run when A is true.


In sbin/attest-enroll line 252:
	rm -f $outdir/tk.pem
              ^-----^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
	rm -f "$outdir"/tk.pem


In sbin/attest-enroll line 373:
ekhash="$(sha256sum $tmp/ek.pub | cut -f1 -d' ' )"
                    ^--^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
ekhash="$(sha256sum "$tmp"/ek.pub | cut -f1 -d' ' )"


In sbin/attest-enroll line 377:
	hostname="$(echo $ekhash | cut -c1-8)"
                         ^-----^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
	hostname="$(echo "$ekhash" | cut -c1-8)"


In sbin/attest-enroll line 404:
	set -- $("$i" "$tmp" "$tmp/ek.pub" "$hostname")
               ^-- SC2046: Quote this to prevent word splitting.

```

osresearch commented:
One additional bit of flexibility that we should consider as part of the pluggable enrollment is that Windows requires specific per-machine file names for the encryption key. I've created a separate issue so we don't forget to think about it after this is merged #129

nicowilliams commented:
> "77 files changed". This is going to take a little while to review...

Er, I can leave out @geoffreythorpe's Mariner commits!

nicowilliams commented:
> The different policies for different keys makes lots of sense. That's something that we should figure out if there is a clean way to specify the alternate policy.

I'm struggling with this :)  I've these ideas (and code for most of them):

 - by `policyDigest` -- not very user-friendly!
 - as a sequence of policy commands -- self-documenting but unwieldy!
 - by name, where the name is an index into a set of hard-coded policies -- not very flexible
 - by name, where the name is a file that contains a `policyDigest` -- not exactly self-documenting!
 - by name, where the name is an executable -- but now we have to execute it, which I think is mostly not a problem, but we have to make sure this is not a name that an untrusted party can give us (well, we have to anyways)
 - by name, where the name is a file that contains a sequence of policy commands that we'd "interpret" -- self-documenting, flexible

Thoughts?

nicowilliams commented:
> Regarding the changes to die(), I agree that having a function inside a library call exit is not the best approach and can make error handling harder. Although having die() not exit means that the name is no longer as appropriate as the Perl-inspired die function. My suggestion is that we split this into five bash functions:

No, it's OK to exit in a function.  I only want to treat it as a function that might not exit because I want to be able to run these functions in an interactive shell and not lose my context on error.  It's just developer convenience.  So `die` will `exit` in all prod cases, so I don't think it needs to be renamed.

nicowilliams commented:
```In sbin/attest-enroll line 34:
. "$BASEDIR/../functions.sh"
  ^------------------------^ SC1090: Can't follow non-constant source. Use a directive to specify location.```

Hmm, but that construction is very useful :(

nicowilliams commented:
```
In sbin/attest-enroll line 142:
	exit ${1:-1}
             ^-----^ SC2086: Double quote to prevent globbing and word splitting.
```

Sure, but that's in `usage`, which is only ever called with `0` or no arguments.  This sort of thing makes me dislike shell-check.

nicowilliams commented:
```
In sbin/attest-enroll line 142:
	exit ${1:-1}
             ^-----^ SC2086: Double quote to prevent globbing and word splitting.
```

Sure, but that's in `usage`, which is only ever called with `0` or no arguments.  This sort of thing makes me dislike shell-check.

nicowilliams commented:
```
In sbin/attest-enroll line 172:
[[ -n ${CONF} && -f ${CONF} ]] && . "${CONF}"
                                    ^-------^ SC1090: Can't follow non-constant source. Use a directive to specify location.
```

This one really has to be like this.

nicowilliams commented:
Also, needing different policies for different secrets really does mean one must use the `send-to-tpm` method, or that one must have a TK per-secret.  Maybe we no longer need the TK method?

nicowilliams commented:
I've rebased and:

 - removed Geoffrey's Mariner patches in order to make this a smaller review
 - push fixes for code review into separate fixup commits (but, reordered, so this was a non-fast-forward push)

nicowilliams commented:
@geoffthorpe: so I've made the `attest-enroll` invocation take an optional directory argument as its second command-line argument, and that would be the directory in which it should do its work.

nicowilliams commented:
This now also handles EKpubs as PEM.

@geoffthorpe: I've also added support for `CHECKOUT` and `COMMIT` hooks.

osresearch commented on Makefile:
```diff
@@ -303,10 +303,12 @@ shellcheck:
 	for file in \
 		sbin/safeboot* \
 		sbin/tpm2-attest \
+		sbin/send-to-tpm.sh \
```
Is there a reason to name these helpers `.sh`?

nicowilliams replied:
There is not.  I'll fix it.

nicowilliams replied:
Fixed.

osresearch commented on functions.sh:
```diff
@@ -114,15 +117,31 @@ tpm2_extend() {
 	tpm2 flushcontext \
 		--transient-object \
-	|| die "tpm2_flushcontext: unable to flush transient handles"
+	|| die "tpm2_flushcontext: unable to flush transient handles" \
```
Seeing any code after `die` confuses my Perl brain, so I think we should have `die` *always* exit, and for these helper functions we should use a separate `error` function (with the `|| error foo || return 1` template).  That way it is clear that `die` will really kill things (and the callers should be checking in case they have any clean up anyway)

nicowilliams replied:
Fixed.

osresearch commented on functions.sh:
```diff
@@ -240,8 +270,8 @@ efivar_write() {
-	efivar_setup "$1"
-	cat "$var" | tail -c +5
+	efivar_setup "${1:-}"
+	tail -c +5 "$var"
```
I think this only *looked* like a "useless use of `cat`" -- the EFI var files are not normal and `tail` tries to `seek()` if its input is not `stdin`, which doesn't always work on `/sys` "files".

nicowilliams replied:
Fixed.

osresearch commented on functions.sh:
```diff
@@ -293,7 +325,162 @@ mount_by_uuid() {
+}
+
+_rand() {
+	if ${USE_TPM2_RAND:-true}; then
```
rather than silently falling back to other techniques if one fails, we should have a default that can be overridden, and if it fails we should report an `error` and return failure.

nicowilliams replied:
Hmm, I mean, I figured try real hard to get some entropy, and then fail.  I can make it not silent, but I figure less configuration == more user-friendly.

osresearch commented on functions.sh:
```diff
@@ -293,7 +325,162 @@ mount_by_uuid() {
+			      -d					\
+			      -nosalt					\
+			      -kfile "$key_file"			\
+			      -iv 00000000000000000000000000000000	\
```
Are there any concerns about reusing the all-zero IV? I'm not a cryptographer, but I noticed that `tpm2 encryptdecrypt` defaults to IV=0 and also warns about it.

nicowilliams replied:
Look carefully!  This is "confounded" AES-256-CBC-HMAC-SHA-256.  Which means that a block of nonce bits is prepended to the plaintext, and then that is encrypted in CBC mode with an IV of all-zeros, but the ciphertext block resulting from encrypting the confounder is then the IV for the first real block of plaintext because of the chaining in CBC.  This isn't strictly necessary -- I could have used an explicit IV, and in earlier versions I did, but that complicated the code a fair bit and the IV still had to be included in the output.  Now the code is much simpler and the only cost is an extra block to encrypt/decrypt.

Confounding is used in Kerberos, and when used without additional chaining between messages it is then practically indistinguishable from an explicit IV -- it just costs an extra cipher block operation.  So, I do have an explicit non-zero IV, in a roundabout way.

osresearch commented on initramfs/bootscript:
```diff
@@ -3,16 +3,23 @@
+
 POLICY_PCR=11
 die() {
+	set +euo pipefail
```
why are the shell flags adjusted here?

nicowilliams replied:
Habit.  I tend to do this in EXIT trap handlers and such to avoid bugs in them causing partial cleanup.  I can remove this.

nicowilliams replied:
Fixed.

osresearch commented on initramfs/bootscript:
```diff
@@ -30,14 +37,62 @@ tar -xvf - -C "$TMPDIR" \
 	. "$STARTUP" || die "startup failed"
 fi
 
+# Authenticated decryption; see functions.sh
```
can we reuse the one from `functions.sh`? We should have it available to the bootscript.

nicowilliams replied:
OK, I wasn't sure.

nicowilliams replied:
Fixed.

osresearch commented on sbin/attest-verify:
```diff
@@ -11,8 +12,13 @@ import subprocess
+db_path = os.environ.get('SAFEBOOT_DB_DIR','build/attest')
+
+tofu_pcrs = []
+if os.path.exists(os.path.join(db_path, 'tofu_pcrs')):
```
If the PCRs are tofu'ed, do they end up getting sync'ed at some point?

nicowilliams replied:
This requires a conversation.  If we're going to use golden PCRs, then how do we decide what they are, especially if there's any reason that they can vary across the plant.  And when firmware gets updated, we need a way to update the golden PCRs.

My preference would be to always have a list of approved firmwares, initramfs and kernel hashes, and allow any PCR 0 and 1 values when they result from extending only with approved values using the eventlog.  But that may not work for all users.

Now, with @geoffthorpe's architecture the attestation servers have a read-only attestation database, so any golden PCR updates would have to go to the enrollment server(s).  And presumably golden PCR updates would be either manual or involve an API invoked by whatever scripts update firmwares.

osresearch commented on sbin/policy.sh:
```diff
@@ -0,0 +1,134 @@
+function usage {
+	((${1:-1} > 0)) && exec 1>&2
+	cat <<EOF
+Usage: $PROG [-o FILE] [-A | -D] POLICY-CMD [ARGS [\\; ...]]
```
over at tpm2-tools, there is a PR that potentially closes my issue https://github.com/tpm2-software/tpm2-tools/issues/2761 and makes it possible to use the policy programs without a real TPM: https://github.com/tpm2-software/tpm2-tools/pull/2762

until then, we should mention that this does talk to the TPM in the usage.

nicowilliams replied:
Nice!  I'll note this.

nicowilliams replied:
Note added.

osresearch commented on sbin/tpm2-attest:
```diff
@@ -319,7 +320,7 @@ attest()
 	|| die "attestation failed"
 
-	unseal < "$TMP/cipher.tgz" \
+	unseal unseal < "$TMP/cipher.tgz" \
```
Doesn't this extra `unseal` trigger `No arguments expected` in `unseal()`?

nicowilliams replied:
Yes, it snuck in by accident and I've removed it.

nicowilliams replied:
Fixed already.

nicowilliams commented:
I should probably add some docs...  Usage messages are nice and all, but not really enough.

nicowilliams commented:
I should probably also teach the `send-to-tpm` and `tpm-receive` scripts to implement the `TK` method.

nicowilliams commented:
And I should make the initramfs bootscript use `tpm-receive`, no?  Currently it duplicates some of that code.

nicowilliams commented:
Also, I should add support for the fix for https://github.com/tpm2-software/tpm2-tools/issues/2779 (not least so I can test it, which I promised I would do).

nicowilliams commented:
Just gotta make sure that `tests/test-enroll.sh` passes again.

nicowilliams commented:
I think this is very close now.

nicowilliams commented:
This now also handles the PEM case using the fix from [`tpm2-tools#2780`](https://github.com/tpm2-software/tpm2-tools/pull/2780).

BTW, it isn't actually necessary for the `TPM2B_PUBLIC` of an `EK` to be correct for the purpose of encrypting to it.  I tried it, and indeed, it doesn't matter if all the attributes or the `policyDigest` of the given `EKpub` match the actual one or not, as the cryptographic name of the target of `TPM2_Duplicate()` and `TPM2_MakeCredential()` is not bound in at all (unlike the cryptographic name of the activation object, which is bound in), so in fact we did not need [`tpm2-tools#2780`](https://github.com/tpm2-software/tpm2-tools/pull/2780), or the hack we had.

I do have a test in `tests/test-enroll.sh` that we do get the same `EKpub` when using PEM as when reading it from the TPM, but we don't actually need that to be the case in GCP and we don't need that to be the case on-prem, so we could just remove that check.  To make that test possible I made `sbin/attest-enroll` fail in the case that an enrollment exists _and_ there was no content to add to it -- this seems a bit silly.  Maybe I should remove all that, as well as the `dd`-the-public-key-over-a-`TPM2B_PUBLIC` hack in `pem2tpm2b_public()`, since we don't need the hashes to match.

osresearch commented on functions.sh:
```diff
@@ -297,3 +311,272 @@ mount_by_uuid() {
+# Compute the policyDigest of a given policy by executing it in a trial
+# session.
+function make_policyDigest {
+	tpm2 flushcontext --transient-object
```
`tpm2_flushsome`?

osresearch commented:
* [ ] rename new tools to match `tpm2-*`
* [ ] yaml front matter for new docs
* [ ] attest enroll text
* [ ] nav page update for attestation
* [ ] document ek policy (I think it is defined by the tcg)
